### ISSUE TITLE
extra_scale is defined wrong

### ISSUE THREAD

**Issue by khle08** [ljanyst/ssd-tensorflow#40]
hi, thanks for your excellent code.

Here are some tiny mistakes when setting extra_scale. Your parameters are 100 times more than the correct parameters.

For SSD_PRESET['vgg300'], extra_scale should be 1.075
For SSD_PRESET['vgg512'], extra_scale should be 1.05

by setting 100 times less than your original number, the anchors would stay inside the original image.

**Comment by ljanyst** [comment]
It looks like you're right. Thanks! I will have a closer look and re-run the training.

**Comment by khle08** [comment]
By the way, The way you generate anchor boxes is by using for loop, which causes so much time on iteration. Here is a better solution to 5 times speed up the whole process using matrix method

`
def get_all_anchors(config, offset=0.5):
    '''
    Compute the default (anchor) boxes based on the given SSD configs.

    It returns the "ratios" of each anchor. By multiplying original
    image size, we can get the real coordinates for each anchor box.
    '''
    result = np.ones((config.num_anchors, 4))
    N = 0

    for i, f_map in enumerate(config.feature_maps):
        '''
        e.g. f_map.size.h, f_map.size.w = (4, 6)
        ctr_h = [[0, 1, 2, 3],        ctr_w = [[0, 0, 0, 0],
                 [0, 1, 2, 3],                 [1, 1, 1, 1],
                 [0, 1, 2, 3],                 [2, 2, 2, 2],
                 [0, 1, 2, 3],                 [3, 3, 3, 3],
                 [0, 1, 2, 3],                 [4, 4, 4, 4],
                 [0, 1, 2, 3]]                 [5, 5, 5, 5]]
        '''
        ctr_h, ctr_w = np.meshgrid(np.arange(f_map.size.h),
                                   np.arange(f_map.size.w))
        ctr_h = (ctr_h.astype(np.float) + offset) / f_map.size.h
        ctr_w = (ctr_w.astype(np.float) + offset) / f_map.size.w

        # Stretch them out and stack them vertically, and transpose x, y dims.
        centers = np.vstack((ctr_h.ravel(),
                             ctr_w.ravel())).transpose()

        # Calculate different sizes along with many aspect ratios.
        sqrt_ratios = np.sqrt(f_map.aspect_ratios)[:, np.newaxis]
        sizes = np.hstack((f_map.scale / sqrt_ratios,
                           f_map.scale * sqrt_ratios))

        if i < len(config.feature_maps) - 1:
            s_prime = np.sqrt(config.feature_maps[i].scale *
                              config.feature_maps[i + 1].scale)
        else:
            s_prime = np.sqrt(config.feature_maps[i].scale *
                              config.extra_scale)

        sizes = np.vstack((sizes, [s_prime, s_prime]))

        # Attach "centers" and "sizes" into predefined "result" array.
        k = centers.shape[0]
        for j in range(sizes.shape[0]):
            result[(N + (j * k)):(N + ((j + 1) * k)), :2] = centers
            result[(N + (j * k)):(N + ((j + 1) * k)), 2:] = sizes[j]

        # Record the last time the array being attached.
        N += centers.shape[0] * sizes.shape[0]

    return result`

Good luck!

**Comment by ljanyst** [comment]
Thanks! Would you mind making a pull request for this later part?

**Comment by khle08** [comment]
But the values returned from your functions are mostly entangling with "namedtuple" module... which is hard to make a partial change.

By using matrix method, the output data would be arranged by array's position.

**Comment by khle08** [comment]
Hi, I have almost finished updating your code. And I will paste the link to you.

But now I'm still not sure if your input labels are as followed:

labels.shape = (batch_size, num_anchors, class_num + 5)
labels = 
                                        < not ratio coordinate >
[[[C1, C2, ... C20, Cbg, ymin, xmin, ymax, xmax],  --
   [C1, C2, ... C20, Cbg, ymin, xmin, ymax, xmax],   |
   ...                                                                               |
   ...                                                                             5776 = 38*38*4
   ...                                                                               |
   ...                                                                               |
   ...                                                                              --
   ...                                                                               |
   ...                                                                             2166 = 19*19*6
   ...                                                                               |
   ...                                                                              --
   ...                                                                               |
   ...                                                                             600 = 10*10*6
   ...                                                                               |
   ...                                                                              --
   ...                                                                             150 = 25*25*6
   ...                                                                              --
   ...                                                                             36 = 3*3*4
   ...                                                                              --
   ...                                                                        ]],  4 = 1*1*4

 [[ ... next batch ...

for each feature map, stacking sequence is:
______38_______              [[C1, C2, ... C20, Cbg, ymin, xmin, ymax, xmax], ... 1
|  -->  1, 2, 3.. |               [C1, C2, ... C20, Cbg, ymin, xmin, ymax, xmax], ... 2
|  --> 39, 40.. | 38  ->   [C1, C2, ... C20, Cbg, ymin, xmin, ymax, xmax], ... 3
|  ...                  |              ...
|______________|               1444 = 38*38
                                        ...                                                                         ]] ... 1444

for those anchors (ymin, xmin, ymax, xmax) that IOU > 0.5, mark them as the label which has highest IOU. And the marking way would be: (for example label = 5)
[0, 0, 0, 0, 0, 1, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] = [C1, C2, ..., C20, Cbg]

for those anchors (ymin, xmin, ymax, xmax) that IOU < 0.5, mark them as background like this:
[0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 1] = [C1, C2, ..., C20, Cbg]

Am I right about your code? Thanks for your patience!
